common/oauthTokenManager.go

```diff
@@ -713,16 +722,135 @@ func (credInfo *OAuthTokenInfo) GetNewTokenFromMSI(ctx context.Context) (*adal.T
+	// Unset the timeout back
+	msiTokenHTTPClient.Timeout = 0
+	return req, resp, err
+}
+
+// isValidArcResponse checks if the key "Www-Authenticate" is unavailable in the header of an http response
+func isValidArcResponse(resp *http.Response) bool {
+	// Parameter for validity is whether "Www-Authenticaite" exists in the response header
+	// "Www-Authenticate" contains the path to the challenge token file for Arc VMs
+	wwwAuthenticateExists := false
+	if resp != nil {
+		if resp.Header != nil {
+			_, wwwAuthenticateExists = resp.Header["Www-Authenticate"]
+		}
+	}
+	return wwwAuthenticateExists
+}
+
+// fixupTokenJson corrects the value of JSON field "not_before" in the Byte slice from blank to a valid value and returns the corrected Byte slice.
+
+// Dated 15th Sep 2021.
+// Token JSON returned by ARC-server endpoint API currently does not set a valid integral value for "not_before" key.
+// If the token JSON already has "not_before" correctly set, this will be a no-op.
+func fixupTokenJson(bytes []byte) []byte {
+	byteSliceToString := string(bytes)
+	separatorString := `"not_before":"`
+	stringSlice := strings.Split(byteSliceToString, separatorString)
+
+	if stringSlice[1][0] != '"' {
+		return bytes
+	}
+
+	// If the value of not_before is blank, set to "now - 5 sec" and return the updated slice
+	notBeforeTimeInteger := uint64(time.Now().Unix() - 5)
+	notBeforeTime := strconv.FormatUint(notBeforeTimeInteger, 10)
+	return []byte(stringSlice[0] + separatorString + notBeforeTime + stringSlice[1])
+}
+
+// GetNewTokenFromMSI gets token from Azure Instance Metadata Service identity endpoint. It first checks if the VM is registered with Azure Arc. Failing that case, it checks if it is an Azure VM.
+// For details, please refer to https://docs.microsoft.com/en-us/azure/active-directory/managed-identities-azure-resources/overview
+// Note: Currently the msiTokenHTTPClient timeout is configured for 30 secs. Should be reduced to 5 sec as IMDS endpoint is local to the machine.
+// Without this change, if some router is configured to not return "ICMP unreachable" then it will take 30 secs to timeout and fallback to ARC.
+// For now, this has been mitigated by checking Arc first, and then Azure
+func (credInfo *OAuthTokenInfo) GetNewTokenFromMSI(ctx context.Context) (*adal.Token, error) {
+	// Try Arc VM
+	req, resp, err := credInfo.queryIMDS(ctx, MSIEndpointArcVM, Resource, IMDSAPIVersionArcVM)
 	if err != nil {
-		return nil, fmt.Errorf("please check whether MSI is enabled on this PC, to enable MSI please refer to https://docs.microsoft.com/en-us/azure/active-directory/managed-identities-azure-resources/qs-configure-portal-windows-vm#enable-system-assigned-identity-on-an-existing-vm. (Error details: %v)", err)
+		// Try Azure VM since there was an error in trying Arc VM
+		reqAzureVM, respAzureVM, errAzureVM := credInfo.queryIMDS(ctx, MSIEndpointAzureVM, Resource, IMDSAPIVersionAzureVM)
+		if errAzureVM != nil {
```

Comment:
a bit of a readability concern here: in general it'd be best to avoid multiple nested if/else statements

Reply:
I tried to reduce the nesting and pushed the code again.